README.md
Plumage-py
==========

Plumage-py is a Python 2.7 implementation of [Plumage](https://github.com/codingatty/Plumage), a module to obtain trademark status information from the United States Patent & Trademark Office's (PTO) [Trademark Status & Document Retrieval](http://tsdr.uspto.gov/) (TSDR) system.  It takes as input either a registration number or application serial number, fetches the corresponding XML data from the PTO's TSDR website, and returns a dictionary of data associated with the specified TSDR entry.

XSL files are sourced from the language-independent sister repository [Plumage](https://github.com/codingatty/Plumage):
* [`ST66.xsl`](https://github.com/codingatty/Plumage/blob/master/ST66.xsl)
* [`ST96.xsl`](https://github.com/codingatty/Plumage/blob/master/ST96.xsl)

For documentation, please see the [Plumage wiki](https://github.com/codingatty/Plumage/wiki).

Plumage (including the [Plumage](https://github.com/codingatty/Plumage), [Plumage-py](https://github.com/codingatty/Plumage-py) and [Plumage-dotnet](https://github.com/codingatty/Plumage-dotnet) repositories) is open source, licensed under the [Apache Software License V2.0](http://www.apache.org/licenses/LICENSE-2.0) (code, including XSL files) and the [Creative Commons Attribution-ShareAlike license V3.0](http://creativecommons.org/licenses/by-sa/3.0/) (documentation). For details, see [License Information](https://github.com/codingatty/Plumage/wiki/License-Information).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.3.0](https://github.com/codingatty/Plumage-py/releases/tag/V1.3.0) - 2018-03-22 (*"Delius"*)
- Support for Python 3; specifically, tested on [Python 2.7.14](https://www.python.org/downloads/release/python-2714/) and [Python 3.6.4](https://www.python.org/downloads/release/python-364/)
- Development environment upgraded to [lxml](http://lxml.de/) 4.2.0

Because the only change in this release is to support Python 3, there is no corresponding release 1.3.0 for
[Plumage-dotnet](https://github.com/codingatty/Plumage-dotnet), which remains at [V1.2.0](https://github.com/codingatty/Plumage-dotnet/releases/tag/V1.2.0).

## [1.2.0](https://github.com/codingatty/Plumage-py/releases/tag/V1.2.0) - 2018-01-28 (*"Cloudbusting"*)
- API change: single-instance values and multiple-instance values are broken into two data fields (see release notes for details)
- XSL relaxation: accept and ignore blank or empty lines
- [Software Package Data Exchange](https://spdx.org/) (SPDX) support
- Upgrade to [Plumage-XSL](https://github.com/codingatty/Plumage) [V1.1.1](https://github.com/codingatty/Plumage/releases/tag/V1.1.1) (for SPDX support)
- Self-test removed in favor of unit testing

## [1.1.0](https://github.com/codingatty/Plumage-py/releases/tag/V1.1.0) - 2016-05-23 (*"The Big Sky"*)
- Support for ST.96 V 2.2.1 (adopted by the US PTO on May 6, 2016)
- Added diagnostic where older ST.96 1-D3 is encountered
- small bug fix from [Plumage V1.0.1](https://github.com/codingatty/Plumage/releases/tag/V1.0.1)

## [1.0.0](https://github.com/codingatty/Plumage-py/releases/tag/V1.0.0) - 2016-05-17 (*"And Dream of Sheep"*)
- Initial formal release of the Python implementation of Plumage. This is essentially the release I've been using for about a year; just formalizing the release.
- Support for ST.66 (release level not identified by USPTO) and ST.96, release 1_D3 


tests/basic.py
import unittest
from testing_context import plumage

class TestUM(unittest.TestCase):

    TESTFILES_DIR = "testfiles\\"
 
    def setUp(self):
        pass

    def test_A001_test_initialize(self):
        '''
        Simple test, just verify TSDRReq can be initialized correctly
        '''
        t = plumage.TSDRReq()
        self.assertFalse(t.XMLDataIsValid)
        self.assertFalse(t.CSVDataIsValid)
        self.assertFalse(t.TSDRData.TSDRMapIsValid)

if __name__ == '__main__':
    unittest.main(verbosity=5)

tests/test_online.py
import sys
import unittest
PYTHON2 = sys.version_info.major == 2
PYTHON3 = sys.version_info.major == 3

from testing_context import plumage

class TestUM(unittest.TestCase):
 
    def setUp(self):
        pass

    def _validate_sample(self, t):
        '''
        Test to confirm content for Python trademark; whether as
        app no. 76/044,902 or ser. no.  2824281
        '''
        self.assertTrue(t.XMLDataIsValid)
        self.assertEqual(t.XMLData[0:55], r'<?xml version="1.0" encoding="UTF-8" standalone="yes"?>')
        self.assertTrue(t.CSVDataIsValid)
        tsdrdata=t.TSDRData
        self.assertTrue(tsdrdata.TSDRMapIsValid)
        self.assertEqual(tsdrdata.TSDRSingle["ApplicationNumber"], "76044902")
        self.assertEqual(tsdrdata.TSDRSingle["ApplicationDate"], "2000-05-09-04:00")
        self.assertEqual(tsdrdata.TSDRSingle["ApplicationDate"][0:10],
                         tsdrdata.TSDRSingle["ApplicationDateTruncated"])
        self.assertEqual(tsdrdata.TSDRSingle["RegistrationNumber"], "2824281")
        if tsdrdata.TSDRSingle["DiagnosticInfoXSLTFormat"] == "ST.96":  #ST.96 does not include time portion in reg date
            self.assertEqual(tsdrdata.TSDRSingle["RegistrationDate"], "2004-03-23")
        else:
            self.assertEqual(tsdrdata.TSDRSingle["RegistrationDate"], "2004-03-23-05:00")

        self.assertEqual(tsdrdata.TSDRSingle["RegistrationDate"][0:10],
                         tsdrdata.TSDRSingle["RegistrationDateTruncated"])
        self.assertEqual(tsdrdata.TSDRSingle["MarkVerbalElementText"], "PYTHON")
        applicant_list = tsdrdata.TSDRMulti["ApplicantList"]
        applicant_info = applicant_list[0]    
        self.assertEqual(applicant_info["ApplicantName"], "PYTHON SOFTWARE FOUNDATION")
        if t.PTOFormat != "ST66":   # non-zipped ST.66 format does not include assignments
            assignment_list = tsdrdata.TSDRMulti["AssignmentList"]
            assignment_0 = assignment_list[0] # Zeroth (most recent) assignment
            self.assertEqual(assignment_0["AssignorEntityName"],
                     "CORPORATION FOR NATIONAL RESEARCH INITIATIVES, INC.")
            self.assertEqual(assignment_0["AssignmentDocumentURL"],
                     "http://assignments.uspto.gov/assignments/assignment-tm-2849-0875.pdf")
        ## Diagnostic info
        self.assertEqual(tsdrdata.TSDRSingle["DiagnosticInfoXSLTURL"],
                         "https://github.com/codingatty/Plumage")
        self.assertEqual(tsdrdata.TSDRSingle["DiagnosticInfoXSLTLicense"],
                         "Apache License, version 2.0 (January 2004)")
        self.assertEqual(tsdrdata.TSDRSingle["DiagnosticInfoXSLTSPDXLicenseIdentifier"],
                         "Apache-2.0")
        self.assertEqual(tsdrdata.TSDRSingle["DiagnosticInfoXSLTLicenseURL"],
                         "http://www.apache.org/licenses/LICENSE-2.0")
        self.assertEqual(tsdrdata.TSDRSingle["DiagnosticInfoImplementationURL"],
                         "https://github.com/codingatty/Plumage-py")
        if PYTHON2: # method renamed from assertRegexpMatches to assertRegex between Py2 and Py3
            self.assertRegexpMatches(tsdrdata.TSDRSingle["DiagnosticInfoImplementationVersion"],
                         r"^\d+\.\d+\.\d+(-(\w+))*$")
        if PYTHON3:
            self.assertRegex(tsdrdata.TSDRSingle["DiagnosticInfoImplementationVersion"],
                         r"^\d+\.\d+\.\d+(-(\w+))*$")
        self.assertEqual(tsdrdata.TSDRSingle["DiagnosticInfoImplementationLicenseURL"],
                         "http://www.apache.org/licenses/LICENSE-2.0")
        self.assertEqual(tsdrdata.TSDRSingle["DiagnosticInfoImplementationLicense"],
                         "Apache License, version 2.0 (January 2004)")
        self.assertEqual(tsdrdata.TSDRSingle["DiagnosticInfoImplementationName"],
                         "Plumage-py")
        self.assertEqual(tsdrdata.TSDRSingle["DiagnosticInfoImplementationSPDXLicenseIdentifier"],
                         "Apache-2.0")

    def test_O001_zipfile_by_serialno(self):
        t = plumage.TSDRReq()
        t.getTSDRInfo("76044902", "s")
        self._validate_sample(t)

    def test_O002_zipfile_by_regno(self):
        t = plumage.TSDRReq()
        t.getTSDRInfo("2824281", "r")
        self._validate_sample(t)

    def test_O003_ST66xmlfile_by_serialno(self):
        t = plumage.TSDRReq()
        t.setPTOFormat("ST66")
        t.getTSDRInfo("76044902", "s")
        self._validate_sample(t)

    def test_O004_ST96xmlfile_by_serialno(self):
        t = plumage.TSDRReq()
        t.setPTOFormat("ST96")
        t.getTSDRInfo("76044902", "s")
        self._validate_sample(t)

    def test_O005_step_by_step(self):
        t = plumage.TSDRReq()
        self.assertFalse(t.XMLDataIsValid)
        self.assertFalse(t.CSVDataIsValid)
        self.assertFalse(t.TSDRData.TSDRMapIsValid)
        t.getXMLData("76044902", "s")
        self.assertTrue(t.XMLDataIsValid)
        self.assertFalse(t.CSVDataIsValid)
        self.assertFalse(t.TSDRData.TSDRMapIsValid)
        t.getCSVData()
        self.assertTrue(t.XMLDataIsValid)
        self.assertTrue(t.CSVDataIsValid)
        self.assertFalse(t.TSDRData.TSDRMapIsValid)
        t.getTSDRData()
        self.assertTrue(t.XMLDataIsValid)
        self.assertTrue(t.CSVDataIsValid)
        self.assertTrue(t.TSDRData.TSDRMapIsValid)
        self._validate_sample(t)
        t.resetXMLData()
        self.assertFalse(t.XMLDataIsValid)
        self.assertFalse(t.CSVDataIsValid)
        self.assertFalse(t.TSDRData.TSDRMapIsValid)

    def test_O099_no_such_app(self):
        '''
        Test no-such-application returns no data, and a Fetch-404 error code
        '''
        t = plumage.TSDRReq()
        t.getTSDRInfo("99999999", "s")
        self.assertFalse(t.XMLDataIsValid)
        self.assertFalse(t.CSVDataIsValid)
        self.assertFalse(t.TSDRData.TSDRMapIsValid)
        self.assertEqual(t.ErrorCode, "Fetch-404") 

if __name__ == '__main__':
    unittest.main(verbosity=5)


SummaryExample01.py
from Plumage import plumage
t = plumage.TSDRReq()
t.getTSDRInfo("2564831", "r")   # get info on reg. no 2,564,831
tsdrdata=t.TSDRData
if tsdrdata.TSDRMapIsValid:
   print "Application serial no: ", tsdrdata.TSDRSingle["ApplicationNumber"]
   print "Trademark text: ", tsdrdata.TSDRSingle["MarkVerbalElementText"]
   print "Application filing date: ", tsdrdata.TSDRSingle["ApplicationDate"]
   print "Registration no: ", tsdrdata.TSDRSingle["RegistrationNumber"]
   # Owner info is in most recent (0th) entry in ApplicantList
   applicant_list = tsdrdata.TSDRMulti["ApplicantList"]
   current_owner_info = applicant_list[0]
   print "Owner:", current_owner_info["ApplicantName"]
   print "Owner address: ", current_owner_info["ApplicantCombinedAddress"]
   # Get most recent event: 0th entry in event list
   event_list = tsdrdata.TSDRMulti["MarkEventList"]
   most_recent_event = event_list[0]
   print "Most recent event: ", most_recent_event["MarkEventDescription"]
   print "Event date: ", most_recent_event["MarkEventDate"]

examples/example01.py
from __future__ import print_function 
from Plumage import plumage
# This example works for Plumage-py v1.2 and above
t = plumage.TSDRReq()
t.getTSDRInfo("2564831", "r")   # get info on reg. no 2,564,831
tsdrdata=t.TSDRData
if tsdrdata.TSDRMapIsValid:
   print("Application serial no: ", tsdrdata.TSDRSingle["ApplicationNumber"])
   print("Trademark text: ", tsdrdata.TSDRSingle["MarkVerbalElementText"])
   print("Application filing date: ", tsdrdata.TSDRSingle["ApplicationDate"])
   print("Registration no: ", tsdrdata.TSDRSingle["RegistrationNumber"])
   # Owner info is in most recent (0th) entry in ApplicantList
   applicant_list = tsdrdata.TSDRMulti["ApplicantList"]
   current_owner_info = applicant_list[0]
   print("Owner:", current_owner_info["ApplicantName"])
   print("Owner address: ", current_owner_info["ApplicantCombinedAddress"])
   # Get most recent event: 0th entry in event list
   event_list = tsdrdata.TSDRMulti["MarkEventList"]
   most_recent_event = event_list[0]
   print("Most recent event: ", most_recent_event["MarkEventDescription"])
   print("Event date: ", most_recent_event["MarkEventDate"])

examples/example01_V02.py
from Plumage import plumage
# This example works for Plumage-py v1.2 (forthcoming)
t = plumage.TSDRReq()
t.getTSDRInfo("2564831", "r")   # get info on reg. no 2,564,831
tsdrdata=t.TSDRData
if tsdrdata.TSDRMapIsValid:
   print "Application serial no: ", tsdrdata.TSDRSingle["ApplicationNumber"]
   print "Trademark text: ", tsdrdata.TSDRSingle["MarkVerbalElementText"]
   print "Application filing date: ", tsdrdata.TSDRSingle["ApplicationDate"]
   print "Registration no: ", tsdrdata.TSDRSingle["RegistrationNumber"]
   # Owner info is in most recent (0th) entry in ApplicantList
   applicant_list = tsdrdata.TSDRMulti["ApplicantList"]
   current_owner_info = applicant_list[0]
   print "Owner:", current_owner_info["ApplicantName"]
   print "Owner address: ", current_owner_info["ApplicantCombinedAddress"]
   # Get most recent event: 0th entry in event list
   event_list = tsdrdata.TSDRMulti["MarkEventList"]
   most_recent_event = event_list[0]
   print "Most recent event: ", most_recent_event["MarkEventDescription"]
   print "Event date: ", most_recent_event["MarkEventDate"]

examples/example01(V01).py
from Plumage import plumage
# This example works through Plumage-py v1.1
t = plumage.TSDRReq()
t.getTSDRInfo("75181334", "s")   # get info on application ser. no 75/181,334
if t.TSDRMapIsValid:
   print "Application serial no: ", t.TSDRMap["ApplicationNumber"]
   print "Trademark text: ", t.TSDRMap["MarkVerbalElementText"]
   print "Application filing date: ", t.TSDRMap["ApplicationDate"]
   print "Registration no: ", t.TSDRMap["RegistrationNumber"]
   # Owner info is in most recent (0th) entry in ApplicantList
   applicant_list = t.TSDRMap["ApplicantList"]
   current_owner_info = applicant_list[0]
   print "Owner:", current_owner_info["ApplicantName"]
   print "Owner address: ", current_owner_info["ApplicantCombinedAddress"]
   # Get most recent event: 0th entry in event list
   event_list = t.TSDRMap["MarkEventList"]
   most_recent_event = event_list[0]
   print "Most recent event: ", most_recent_event["MarkEventDescription"]
   print "Event date: ", most_recent_event["MarkEventDate"]

Plumage/__init__.py
'''
Obtain information on a federal trademark registration or application
for registration from the US Patent and Trademark Office's Trademark
Status and Document Retrieval (TSDR) system at http://tsdr.uspto.gov.

PREREQUISITES
lxml XML toolkit, http://lxml.de/

EXAMPLE

>>> from Plumage import plumage
>>> t = plumage.TSDRReq()
>>> # Python trademark application, ser. no. 76/044,902
>>> t.getTSDRInfo("s", "76044902")
>>> # Show application number and date, registration number and date...
>>> t.tmdict["ApplicationNumber"]
'76044902'
>>> t.tmdict["ApplicationDate"]
'2000-05-09-04:00'
>>> t.tmdict["RegistrationNumber"]
'2824281'
>>> t.tmdict["RegistrationDate"]
'2004-03-23-05:00'
>>> # Show trademark text and owner...
>>> t.tmdict["MarkVerbalElementText"]
'PYTHON'
>>> # Show current status of mark, and date of that status...
>>> t.tmdict["MarkCurrentStatusExternalDescriptionText"]
'A Sections 8 and 15 combined declaration has been accepted and acknowledged.'
>>> t.tmdict["MarkCurrentStatusDate"]
'2010-09-08-04:00'
>>> applicant_info = t.TSDRMap["ApplicantList"][0]
>>> applicant_info["ApplicantName"]
'PYTHON SOFTWARE FOUNDATION'
>>>
'''
